Introduce waitTipChanged() mining interface, replace RPCNotifyBlockChange, drop CRPCSignals & g_best_block

Sjors opened the pull request:
This continues the work in #30200 so that a future Stratum v2 Template Provider (see #29432) can avoid accessing node internals. It needs to know when a new block arrives in order to push new templates to connected clients.

`waitTipChanged()` uses a new kernel notification `notifications().m_tip_block_mutex`, which this PR also introduces (a previous version used `g_best_block`).

In order to ensure the new method works as intended, the `waitfornewblock`, `waitforblock` and `waitforblockheight` RPC methods are refactored to use it. This allows removing `RPCNotifyBlockChange`.

There's a commit to add (direct) tests for the methods that are about to be refactored:
- `waitfornewblock` was already implicitly tested by `feature_shutdown.py`.
- `waitforblockheight` by `feature_coinstatsindex.py` and `example_test.py`

This PR renames `getTipHash()` to `getTip()` and returns a `BlockRef` (renamed from `BlockKey`) so that callers can use either the height or hash.

The later commits make trivial improvements to the `waitfor*` RPC calls (not needed for this PR).

The `waitTipChanged()` method could probably also be used for the longpoll functionality in `getblocktemplate`, but I'm a bit reluctant to touch that.

`RPCServer::OnStarted` no longer does anything and `RPCServer::OnStopped` merely prints a log statement. They were added in #5711 as a refactor. This PR drops them entirely.

Finally `g_best_block` is also dropped.

Sjors committed:
rpc: add test for waitforblock and waitfornewblock

Sjors committed:
rpc: fix waitfornewblock description

The waitforblock RPC method takes a hash argument and waits for that specific block.  The waitfornewblock waits for any new block. This commit fixes the documentation.

Sjors committed:
rpc: check for negative timeout arg in waitfor*

Sjors committed:
Rename getTipHash() to getTip() and return BlockRef

Sjors commented:
Some of the functional tests that use the affected RPC calls still fail under TSAN, so keeping this draft until that's resolved.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -860,6 +860,25 @@ class MinerImpl : public Mining
+        {
+            WAIT_LOCK(g_best_block_mutex, lock);
+            while (!chainman().m_interrupt && std::chrono::steady_clock::now() < deadline) {
+                auto check_time = std::chrono::steady_clock::now() + std::min(timeout, MillisecondsDouble(1000));
```
I changed this in the last push, previously it would always wait for one second each "round" even if the deadline was shorter.

Sjors commented:
Rebased after #30356 (`#include` order).

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -861,6 +861,25 @@ class MinerImpl : public Mining
+            while (!chainman().m_interrupt && std::chrono::steady_clock::now() < deadline) {
+                auto check_time = std::chrono::steady_clock::now() + std::min(timeout, MillisecondsDouble(1000));
+                g_best_block_cv.wait_until(lock, check_time);
+                if (g_best_block != previous_hash) break;
```
In commit "Add waitTipChanged to Mining interface" (4554b3b1adf678e17d479d72738f4f2089971c4a)

If you adopt suggestion above for caller to pass in the previous_hash value, you will want to delete this line and add `g_best_block == previous_hash &&` to the while loop condition so this function does not block if the hash is already different when it is called.

Sjors replied:
Done

ryanofsky commented on test/functional/rpc_blockchain.py:
```diff
@@ -505,6 +513,56 @@ def _test_stopatheight(self):
+        node.invalidateblock(b2.hash)
+        node.reconsiderblock(b2.hash)
+
+        assert_equal(
+            node.waitfornewblock(timeout=2)['hash'],
+            current_hash)
```
In commit "rpc: add test for waitforblock and waitfornewblock" (a18cb7adf192bae481c879fe865f1c159b2b8622)

I don't think I understand what this test is checking for. Maybe there could be a comment that describes what the test setup is doing what a failed result might look like. It seems like it is just creating a small fork and then invalidating the fork tip, so it shouldn't affect the most work tip or current chain. I would expect to see wait_until call after the invalidateblock call which verifies it has the intended effect. Also the 2 millisecond timeout seems small so if this is supposed to wait for something it doesn't seem like it waiting very long.

Same comments also apply to the  _test_waitforblock test below, and maybe it would be good to dedupe the test setup shared by these functions, moving it into a helper method they both could call.

Sjors replied:
I don't fully understand the original test either. Will look into it later.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -255,7 +255,7 @@ void RPCNotifyBlockChange(const CBlockIndex* pindex)
 {
     return RPCHelpMan{"waitfornewblock",
-                "\nWaits for a specific new block and returns useful info about it.\n"
+                "\nWaits for any new block and returns useful info about it.\n"
```
In commit "rpc: fix waitfornewblock description" (e2a1347db07933d1c6e40ab45f2ca4c56c691fed)

This is a good documentation fix, but as described previously this waitfornewblock interface is inherently racy. Ideally it would accept an `current_block` option or argument, and return as soon as the tip is not not pointing to that block. This could be something to add in a followup PR, though, because it is not directly relevant to this one.

Sjors replied:
I made a note to either add that argument in this PR or make a followup.

ryanofsky commented on src/init.cpp:
```diff
@@ -427,17 +427,12 @@ static void registerSignalHandler(int signal, void(*handler)(int))
 {
-    rpc_notify_block_change_connection.disconnect();
-    RPCNotifyBlockChange(nullptr);
-    g_best_block_cv.notify_all();
```
In commit "Replace RPCNotifyBlockChange with waitTipChanged()" (9d3b07526110a842c9592425b89785e688a8f46c)

g_best_block_cv still exists, so I don't think you should delete this line

Can move it to the place this function is called when this function is deleted in the next commit.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -324,14 +314,19 @@ static RPCHelpMan waitforblock()
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while(IsRPCRunning() && (timeout == 0 || std::chrono::steady_clock::now() < deadline)) {
+        if (block.hash == hash) break;
+        auto res = timeout ? miner.waitTipChanged(std::chrono::milliseconds(timeout)) : miner.waitTipChanged();
```
In commit "Replace RPCNotifyBlockChange with waitTipChanged()" (9d3b07526110a842c9592425b89785e688a8f46c)

Should pass `block.hash` as `waitTipChanged` argument here to avoid race condition (see earlier suggestion)

Sjors replied:
Done

ryanofsky reviewed (commented):
Code review e367258c9b37f0494bdc9d36a6a48adb62f9c94f. This is a nice change, but I didn't ack because I think it would be a regression to not signal g_best_block_cv on shutdown. I posted other suggestions to consider too but they are not critical.

Sjors committed:
refactor: rename BlockKey to BlockRef

Sjors committed:
Replace RPCNotifyBlockChange with waitTipChanged()

This refactoring commit uses the newly introduced waitTipChanged mining interface method to replace the RPCNotifyBlockChange mechanism.

Sjors commented:
Rebased and restored signalling `g_best_block_cv` on shutdown.

I created `interfaces/types.h`, moved `BlockKey` there and renamed it to `BlockRef`. I then renamed `getTipHash()` to `getTip()` instead of adding `getTipHeight()`. `waitTipChanged` also returns a `BlockKey`.

Addressed inline comments, except that I still need to look into the tests.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -39,6 +40,15 @@ class Mining
+     * @returns hash and height for the new tip or the previous tip if
+     *          interrupted or after the timeout
+     */
+    virtual std::pair<uint256, int> waitTipChanged(MillisecondsDouble timeout = MillisecondsDouble::max()) = 0;
```
`getTip()` and `waitTipChanged()` now both return `BlockRef`.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -861,6 +861,25 @@ class MinerImpl : public Mining
 
+    std::pair<uint256, int> waitTipChanged(MillisecondsDouble timeout) override
+    {
+        uint256 previous_hash{WITH_LOCK(::cs_main, return chainman().ActiveChain().Tip()->GetBlockHash();)};
```
Done. I also simplified the `waitfornewblock` (etc) RPC call to just call getTip() and then pass that into `waitTipChanged`. That gets rid of `struct CUpdatedBlock`.

Sjors commented:
As for race conditions with RPC calls that rely on `waitForBlock()`: 

* `waitforblock [hash]`: the while loop breaks if `block.hash == hash`, which together with the early return inside `waitTipChanged` should prevent any blockage
* `waitforblockheight`: similar to `waitforblock`, the while loop checks `block.height >= height`  
* `waitfornewblock`: I opened a draft PR to add an optional `blockhash` argument #30635

Sjors commented:
I simplified the `waitforblock` and `waitfornewblock` tests.

They're not very thorough at the moment, because `reconsiderblock` is either atomic or too fast. A better way to test the asynchronous behaviour would be to spin up a second node, and feed it blocks through `add_p2p_connection`.

ryanofsky reviewed (commented):
Code review dcf984b055a367facf7704eb8055619d6fe64a55. I think there might be an infinite loop bug if waitfornewblock is called with timeout of 0, and it looks like there are some other quirks. Left a few suggestions below.

Sjors committed:
rpc: recommend -rpcclienttimeout=0 for waitfor*

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -372,15 +354,16 @@ static RPCHelpMan waitforblockheight()
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while(IsRPCRunning() && (timeout == 0 || std::chrono::steady_clock::now() < deadline)) {
+        if (block.height >= height) break;
+        block = timeout ? miner.waitTipChanged(block, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block);
```
Fixed (in the same way as `waitforblock`)

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -324,14 +306,14 @@ static RPCHelpMan waitforblock()
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while(IsRPCRunning() && (timeout == 0 || std::chrono::steady_clock::now() < deadline)) {
+        if (block.hash == hash) break;
+        block = timeout ? miner.waitTipChanged(block, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block);
```
Fixed, though in a slightly different way.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -36,8 +38,18 @@ class Mining
+     * @returns hash and height for the new tip or the previous tip if
+     *          interrupted or after the timeout
+     */
+    virtual BlockRef waitTipChanged(BlockRef current_tip, MillisecondsDouble timeout = MillisecondsDouble::max()) = 0;
```
Done

Sjors commented:
Re https://github.com/bitcoin/bitcoin/pull/30409#discussion_r1713879417

> it looks like this will wait 1000ms+1000ms instead of 1000ms+500ms

Indeed it seems to pick the wrong wait time after the first inner loop. If fixed that and the overflow, though in a slightly different way than the AI overlord suggested.

I also added a `g_best_block == uint256()` check because it starts 0 until the first `UpdateTip()`.  

Changed current_tip argument to `uint256` instead of `BlockRef`.

I also added a commit to recommend `-rpcclienttimeout=0` for these three calls.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -870,6 +870,30 @@ class MinerImpl : public Mining
+            while ((g_best_block == uint256() || g_best_block == current_tip) && !chainman().m_interrupt) {
+                now = std::chrono::steady_clock::now();
+                if (now >= deadline) break;
+                const MillisecondsDouble remaining{deadline - now};
+                const auto interval{std::min(remaining, tick)};
+                g_best_block_cv.wait_until(lock, now + interval);
```
In commit "Add waitTipChanged to Mining interface" (d1c197fbbad86e3f966d9bf9abe4dd0aa76bc83d)

Not important but I think it would be clearer to condense these 3 lines to just:

```c++
g_best_block_cv.wait_until(lock, std::min(deadline, now + tick));
```

I don't think it helps to introduce 2 extra variables and subtract `now` then add `now` back.

Sjors replied:
Indeed that's nicer.

ryanofsky reviewed (approved):
Code review ACK 9ba60b3a2c22dcd1c9e7b4880f0adb678e4e8d63. Just fixed a few bugs to make waiting more reliable since last review and changed test logging. I like the new change to check for overflows generally in waitTipChanged, instead of just checking specifically for the max timeout value. That was better than the suggestion.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -40,6 +41,16 @@ class Mining
+    /**
+     * Waits for the tip to change
+     *
+     * @param[in] current_tip the current tip, in case of a race condition
```
Taken (mostly).

Sjors commented on src/interfaces/mining.h:
```diff
@@ -40,6 +41,16 @@ class Mining
+     *
+     * @param[in] current_tip the current tip, in case of a race condition
+     * @param[in] timeout how long to wait for a new tip
+     * @returns hash and height for the new tip or the previous tip if
+     *          interrupted or after the timeout
```
Done

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -326,14 +306,21 @@ static RPCHelpMan waitforblock()
+    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while (IsRPCRunning()) {
+        auto now{std::chrono::steady_clock::now()};
```
In commit "Replace RPCNotifyBlockChange with waitTipChanged()" (bf3377f4bf988f2477944ca54c4dee33b9b45b6e)

`now` variable is unused if timeout is zero, so this line could be moved inside the `if (timeout)` block below.

Sjors replied:
Done here and in `waitforblockheight`.

ryanofsky reviewed (approved):
Code review ACK d880419cb65fd9d446057ea05384dad55da0f292. Only changes since list review are documentation improvements and a small code simplification.

Overall this PR makes nice improvements to the waitfor* RPCs and simplifies the code getting rid of boost signals and a global variable.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -326,14 +306,21 @@ static RPCHelpMan waitforblock()
+
+    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while (IsRPCRunning()) {
+        if (block.hash == hash) break;
```
In commit "Replace RPCNotifyBlockChange with waitTipChanged()" (b4d37fa70506129ba4a4d218bc49ce8ca9d8380c)

Not important, but could simplify these two lines to `while (IsRPCRunning() && block.hash != hash)`

Similarly, could simplify to `while (IsRPCRunning() && block.height < height)` in waitforblockheight below.

Sjors replied:
Thanks, will do if I have to retouch.

Sjors replied:
Done

ryanofsky reviewed (approved):
Code review ACK e8176694bb5508b2b488a8c79a6001d2fa625302, just reduced a variable scope since last review.

TheCharlatan commented on src/rpc/blockchain.cpp:
```diff
@@ -278,16 +260,14 @@ static RPCHelpMan waitfornewblock()
+
+    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (IsRPCRunning()) {
+        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
```
In commit b4d37fa70506129ba4a4d218bc49ce8ca9d8380c

I think there is a significant change in behaviour here between `waitTipChanged`, or rather `g_best_block`, and the rpc notification through `blockTip`. AFAICT the `blockTip` notification reports on completed operations in `ActivateBestChain` and `InvalidatedBlock`, while `g_best_block` is updated through the lower-level `ConnectTip` and `DisconnectTip`, which may respectively be called iteratively by `ActivateBestChain` and `InvalidateBlock`. So for operations involving multiple blocks this might yield different results here. I'm not sure if these RPCs should report on these more ephemeral state changes rather than the results of the respective operations.

I have not actually tested this, so I might be mistaken here, in which case I think it would still be good to have a note in the commit message comparing the two methods.

Sjors replied:
`g_best_block_cv.notify_all();` is called by `Chainstate::UpdateTip()`, which indeed is called by `ConnectTip` and `DisconnectTip()`. Still wrapping my head around what the previous approach was. I doubt these RPC's were used outside of tests, but it would make sense to document the different.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30409#discussion_r1727105521

This is a good catch, but I"m not sure there is a observable change of behavior in `ActivateBestChain` because cs_main is held the whole time `blockTip` / `uiInterface.NotifyBlockTip` / `RPCNotifyBlockChange` is called and while the old `latestblock` global (now replaced by `g_best_block`) was updated:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3552

So I don't think `waitTipChanged` will actually return during any intermediate tip changes, because it can't return until it locks `cs_main`, so it should just report the latest block as before.

However, `InvalidBlock` seems like a different story because it is releasing cs_main between disconnecting blocks:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3701

and it would previously only notify about a tip change and update `latestblock` at the end:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3783

But now it will update `g_best_block` and release `cs_main` while it is disconnecting blocks. This seems like something worth documenting, but probably unlikely to cause any problems.

TheCharlatan replied:
Couple of months ago I was thinking how to get rid of `g_best_block` and its friends entirely in the context of the kernel library. We already have two interfaces for notifying if something changed during validation, having a third one, that is also globally mutable and potentially dangerous if multiple chainstate managers are used is not ideal. I did not make progress then, but looking at it again now, I think this might be the opportunity to get rid of it.

How about duplicating the best block logic into the kernel notifications like I've done here https://github.com/TheCharlatan/bitcoin/commit/3c665065f132fd80ea3b1733785b98e45c8b371e (our locking conventions still puzzle me a bit, maybe there is a better way to do it)? This would avoid any change in behavior altogether and mostly make this a refactor. In a follow-up the remaining users of `g_best_block` could then be migrated and the global deleted.

I think the current user of `g_best_block` (getblocktemplate() in mining.cpp), could be migrated to this interface as well. I don't really see the downside of not being notified on these interim states, since it is undesirable to build work on top of them.

Sjors replied:
@TheCharlatan do you want to make a separate PR for the kernel notification changes, or should I just include them here?

TheCharlatan replied:
I think it would be good to integrate them here, and I'll try and follow up with removing the `m_best_header` global from validation.

ryanofsky replied:
If we are going to add a new `m_tip_block` variable to replace `g_best_block` seems like it would be good to make replacement complete and eliminate `g_best_block`. I think following could work (compiles but untested):

```diff
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -739,7 +739,6 @@ static RPCHelpMan getblocktemplate()
     {
         // Wait to respond until either the best block changes, OR a minute has passed and there are more transactions
         uint256 hashWatchedChain;
-        std::chrono::steady_clock::time_point checktxtime;
         unsigned int nTransactionsUpdatedLastLP;
 
         if (lpval.isStr())
@@ -760,19 +759,14 @@ static RPCHelpMan getblocktemplate()
         // Release lock while waiting
         LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
-
-            WAIT_LOCK(g_best_block_mutex, lock);
-            while (g_best_block == hashWatchedChain && IsRPCRunning())
-            {
-                if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
-                {
-                    // Timeout: Check transactions for update
-                    // without holding the mempool lock to avoid deadlocks
-                    if (miner.getTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += std::chrono::seconds(10);
-                }
+            MillisecondsDouble checktxtime{std::chrono::minutes(1)};
+            while (tip == hashWatchedChain && IsRPCRunning()) {
+                tip = miner.waitTipChanged(hashWatchedChain, checktxtime).hash;
+                // Timeout: Check transactions for update
+                // without holding the mempool lock to avoid deadlocks
+                if (miner.getTransactionsUpdated() != nTransactionsUpdatedLastLP)
+                    break;
+                checktxtime = std::chrono::seconds(10);
             }
         }
         ENTER_CRITICAL_SECTION(cs_main);
--- a/src/rpc/server.cpp
+++ b/src/rpc/server.cpp
@@ -303,7 +303,6 @@ void StopRPC(const std::any& context)
         LogPrint(BCLog::RPC, "Stopping RPC\n");
         WITH_LOCK(g_deadline_timers_mutex, deadlineTimers.clear());
         DeleteAuthCookie();
-        g_best_block_cv.notify_all();
         Assert(EnsureAnyNodeContext(context).notifications)->m_tip_block_cv.notify_all();
         LogPrint(BCLog::RPC, "RPC stopped.\n");
     });
--- a/src/test/validation_chainstate_tests.cpp
+++ b/src/test/validation_chainstate_tests.cpp
@@ -4,6 +4,7 @@
 //
 #include <chainparams.h>
 #include <consensus/validation.h>
+#include <node/kernel_notifications.h>
 #include <random.h>
 #include <rpc/blockchain.h>
 #include <sync.h>
@@ -69,14 +70,14 @@ BOOST_AUTO_TEST_CASE(validation_chainstate_resize_caches)
 BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
 {
     ChainstateManager& chainman = *Assert(m_node.chainman);
-    uint256 curr_tip = ::g_best_block;
+    uint256 curr_tip = m_node.notifications->m_tip_block;
 
     // Mine 10 more blocks, putting at us height 110 where a valid assumeutxo value can
     // be found.
     mineBlocks(10);
 
     // After adding some blocks to the tip, best block should have changed.
-    BOOST_CHECK(::g_best_block != curr_tip);
+    BOOST_CHECK(m_node.notifications->m_tip_block != curr_tip);
 
     // Grab block 1 from disk; we'll add it to the background chain later.
     std::shared_ptr<CBlock> pblockone = std::make_shared<CBlock>();
@@ -91,15 +92,15 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
     // Ensure our active chain is the snapshot chainstate.
     BOOST_CHECK(WITH_LOCK(::cs_main, return chainman.IsSnapshotActive()));
 
-    curr_tip = ::g_best_block;
+    curr_tip = m_node.notifications->m_tip_block;
 
     // Mine a new block on top of the activated snapshot chainstate.
     mineBlocks(1);  // Defined in TestChain100Setup.
 
     // After adding some blocks to the snapshot tip, best block should have changed.
-    BOOST_CHECK(::g_best_block != curr_tip);
+    BOOST_CHECK(m_node.notifications->m_tip_block != curr_tip);
 
-    curr_tip = ::g_best_block;
+    curr_tip = m_node.notifications->m_tip_block;
 
     BOOST_CHECK_EQUAL(chainman.GetAll().size(), 2);
 
@@ -138,7 +139,7 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
     // g_best_block should be unchanged after adding a block to the background
     // validation chain.
     BOOST_CHECK(block_added);
-    BOOST_CHECK_EQUAL(curr_tip, ::g_best_block);
+    BOOST_CHECK_EQUAL(curr_tip, m_node.notifications->m_tip_block);
 }
 
 BOOST_AUTO_TEST_SUITE_END()
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -107,10 +107,6 @@ const std::vector<std::string> CHECKLEVEL_DOC {
  * */
 static constexpr int PRUNE_LOCK_BUFFER{10};
 
-GlobalMutex g_best_block_mutex;
-std::condition_variable g_best_block_cv;
-uint256 g_best_block;
-
 const CBlockIndex* Chainstate::FindForkInGlobalIndex(const CBlockLocator& locator) const
 {
     AssertLockHeld(cs_main);
@@ -2987,12 +2983,6 @@ void Chainstate::UpdateTip(const CBlockIndex* pindexNew)
         m_mempool->AddTransactionsUpdated(1);
     }
 
-    {
-        LOCK(g_best_block_mutex);
-        g_best_block = pindexNew->GetBlockHash();
-        g_best_block_cv.notify_all();
-    }
-
     std::vector<bilingual_str> warning_messages;
     if (!m_chainman.IsInitialBlockDownload()) {
         const CBlockIndex* pindex = pindexNew;
--- a/src/validation.h
+++ b/src/validation.h
@@ -85,11 +85,6 @@ enum class SynchronizationState {
     POST_INIT
 };
 
-extern GlobalMutex g_best_block_mutex;
-extern std::condition_variable g_best_block_cv;
-/** Used to notify getblocktemplate RPC of new tips. */
-extern uint256 g_best_block;
-
 /** Documentation for argument 'checklevel'. */
 extern const std::vector<std::string> CHECKLEVEL_DOC;
 
```

TheCharlatan replied:
Thanks ryanofsky, that is exactly what I had in mind for a potential followup - happy to have this in one go. I still need to think through the trade-offs for the block template rpc, but I think this might actually be better.

Sjors committed:
Add waitTipChanged to Mining interface

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

Sjors commented:
I incorporated https://github.com/TheCharlatan/bitcoin/commit/3c665065f132fd80ea3b1733785b98e45c8b371e so the `waitTipChanged()` implementation uses the kernel signal instead of `g_best_block`. Also rebased.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -870,6 +871,27 @@ class MinerImpl : public Mining
+            while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
+                now = std::chrono::steady_clock::now();
+                if (now >= deadline) break;
+                notifications().m_tip_block_cv.wait_until(lock, std::min(deadline, now + tick));
```
In commit "Add waitTipChanged to Mining interface" (078fc60f198eb7a3bcccb557416fe1d4498f26e7)

One problem with switching from g_best_block_cv to m_tip_block_cv is that this m_tip_block_cv is not signaled when StopRPC is called, which could mean shutdown hangs for a second instead of being instantaneous. I believe this could be fixed in earlier commit 8400d49ffec47add2dc27267ae13ede5d2fabeea with:

```diff
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -282,7 +282,7 @@ void Shutdown(NodeContext& node)
 
     StopHTTPRPC();
     StopREST();
-    StopRPC();
+    StopRPC(&node);
     StopHTTPServer();
     for (const auto& client : node.chain_clients) {
         client->flush();
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -139,7 +139,7 @@ public:
         // Stop RPC for clean shutdown if any of waitfor* commands is executed.
         if (args().GetBoolArg("-server", false)) {
             InterruptRPC();
-            StopRPC();
+            StopRPC(m_context);
         }
     }
     bool shutdownRequested() override { return ShutdownRequested(*Assert(m_context)); };
--- a/src/rpc/server.cpp
+++ b/src/rpc/server.cpp
@@ -11,6 +11,7 @@
 #include <common/system.h>
 #include <logging.h>
 #include <node/context.h>
+#include <node/kernel_notifications.h>
 #include <rpc/server_util.h>
 #include <rpc/util.h>
 #include <sync.h>
@@ -293,16 +294,17 @@ void InterruptRPC()
     });
 }
 
-void StopRPC()
+void StopRPC(const std::any& context)
 {
     static std::once_flag g_rpc_stop_flag;
     // This function could be called twice if the GUI has been started with -server=1.
     assert(!g_rpc_running);
-    std::call_once(g_rpc_stop_flag, []() {
+    std::call_once(g_rpc_stop_flag, [&]() {
         LogPrint(BCLog::RPC, "Stopping RPC\n");
         WITH_LOCK(g_deadline_timers_mutex, deadlineTimers.clear());
         DeleteAuthCookie();
         g_best_block_cv.notify_all();
+        Assert(EnsureAnyNodeContext(context).notifications)->m_tip_block_cv.notify_all();
         LogPrint(BCLog::RPC, "RPC stopped.\n");
     });
 }
--- a/src/rpc/server.h
+++ b/src/rpc/server.h
@@ -9,6 +9,7 @@
 #include <rpc/request.h>
 #include <rpc/util.h>
 
+#include <any>
 #include <functional>
 #include <map>
 #include <stdint.h>
@@ -172,7 +173,7 @@ extern CRPCTable tableRPC;
 
 void StartRPC();
 void InterruptRPC();
-void StopRPC();
+void StopRPC(const std::any& context);
 UniValue JSONRPCExec(const JSONRPCRequest& jreq, bool catch_errors);
 
 #endif // BITCOIN_RPC_SERVER_H
```

ryanofsky reviewed (commented):
Code review 93176016322687f3ad9ade8f32c1ff1392ee6470, but I'm concerned code is failing to signal m_tip_block_cv on shutdown so it could make shutdown slower. Suggested a fix below. It also seems like it would be easy to eliminate g_best_block completely, so suggested a way to do that too.

Sjors committed:
node: Track last block that received a blockTip notification

Also signal m_tip_block_cv when StopRPC is called, for
consistency with g_best_block_cv. This is handled in
StopRPC instead of OnRPCStopped() because the latter
is deleted in a later commit.

Co-authored-by: TheCharlatan <seb.kung@gmail.com>
Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors committed:
Remove unused CRPCSignals

They are no longer used for anything since RPCNotifyBlockChange was replaced with waitTipChanged() from the mining interface.

ryanofsky committed:
Remove unused g_best_block

ryanofsky committed:
rpc: use waitTipChanged for longpoll

This removes the last remaining use of g_best_block by the RPC.

Sjors commented:
Alright then, dropping `g_best_block` as well!

Some tests failed with `EnsureAnyNodeContext(context).notifications`, so I fixed that (relative to the above patches). In `init.cpp` the RPC server starts at step 4a, but the notifications are added at step 7. This case is hit by several functional tests that shut the node down early.

Sjors commented:
Rebased to ensure it still works with CMake. This PR only adds a header file (`interfaces/types.h`) which don't need to be added to `CMakeLists.txt`. If #30664 lands before this PR, I'll have to drop the addition to `src/Makefile.am`.

Sjors commented on src/node/kernel_notifications.cpp:
```diff
@@ -50,6 +50,12 @@ namespace node {
 {
+    {
+        LOCK(m_tip_block_mutex);
+        m_tip_block= index.GetBlockHash();
```
Fixed

Sjors commented:
Based on IRC discussion today it appears there was no need to keep `src/Makefile.am`, so I dropped that: 1b2cdef410abb7fa900c52809ebca249cdaa099c -> 02272716dce292618453d0407988118bc5f07140

Sjors commented:
Rebased after #8230.

TheCharlatan commented:
> Rebased after https://github.com/bitcoin/bitcoin/pull/8230.

Huh, that was merged 8 years ago? :smile:

Sjors commented:
@TheCharlatan oops, I meant #30750.

Sjors commented:
Rebased after trivial conflict with #29553.

ryanofsky reviewed (approved):
Code review ACK 6c419285720fe5a954e01f09cbe053a0f51bef47. Since last review just rebased, added suggested m_tip_block_cv notify call on shutdown, and added two new commits dropping g_best_block.

sedited reviewed (approved):
ACK 6c419285720fe5a954e01f09cbe053a0f51bef47

But I think it would be good to get more review here. I am fairly convinced that the changes to when the notifications are issued now are good, but the change in logic is a bit tricky to follow.

Sjors commented:
Paging some people who touched this code in recent years... well, decades: @sipa, @theuni

Sjors commented:
> the change in logic is a bit tricky to follow.

Any particular commit that's problematic? I could try splitting it.

TheCharlatan commented:
> Any particular commit that's problematic? I could try splitting it.

I don't think so. The logic in the changed lines is relatively easy to understand, but it takes a bit of time to check where the notifications are issued and what the differences between the two might be. I don't think dividing anything here might help with that.

ryanofsky commented:
> > the change in logic is a bit tricky to follow.
> 
> Any particular commit that's problematic? I could try splitting it.

I think the PR is good in it's current form, but if I wanted to be more confident it wasn't changing behavior I would not change code from using `g_best_block` and `::latestblock` directly to calling `waitTipChanged()`. Instead, I would first remove `g_best_block` and replace it with `KernelNotifications::m_tip_block`, then I would remove `::latestblock` and replace it with `KernelNotifications::m_tip_block`, then finally introduce `waitTipChanged()` and change RPCs to call `waitTipChanged()` instead of using `m_tip_block` directly.

The drawback to this approach is it would change the same code twice times, and increase the amount of code reviewers need to look at and intermediate state they need to think about, which is the same problem as the approach I took in #24230. So I wouldn't necessarily recommend that. (For #24230 I think I will simplify it and keeping the granular commits in a side branch with the same overall diff. So if you want to just review the changes for correctness you can look at the PR and if you want to be more sure behavior isn't changing you can look at the side branch.)

Sjors commented:
Trivial rebase after #30509.

ryanofsky reviewed (approved):
Code review ACK 966027bdda53d150321af7db48b57f9756c54e68, no changes since last rebase

sedited reviewed (approved):
Re-ACK 966027bdda53d150321af7db48b57f9756c54e68

Sjors commented:
Trivial `#include` rebase after #30546.

ryanofsky reviewed (approved):
Code review ACK 76ec301c60e2530de7f58e2e4cea3d15c6a77cbc. Just simple rebase since last review

Sjors commented:
Rebased after #30440. Just some `#include` and `using` conflicts since that PR also touched the mining interface.

ryanofsky reviewed (approved):
Code review ACK 7942951e3fcc58f7db0059546d03be9ea243f1db. Just rebased since last review

sedited reviewed (approved):
Re-ACK 7942951e3fcc58f7db0059546d03be9ea243f1db

maflcko reviewed (commented):
review ACK 7942951e3fcc58f7db0059546d03be9ea243f1db 🛋

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 7942951e3fcc58f7db0059546d03be9ea243f1db 🛋
KRzS4gS3SXCvu+SqXrmOlO59h7EfpDm03wNYXbXyxcUpf1uSvRs0Wxsdg7n96gG9dwQPet+ZPAL6FR2cNKqBCw==
```

</details>

Sjors commented on src/rpc/server.cpp:
```diff
@@ -311,16 +312,19 @@ void InterruptRPC()
         WITH_LOCK(g_deadline_timers_mutex, deadlineTimers.clear());
         DeleteAuthCookie();
         g_rpcSignals.Stopped();
+        node::NodeContext& node = EnsureAnyNodeContext(context);
```
I probably just copied similar code from another place :-)

Can make a followup, though `*Assert(AnyPtr(context))` looks pretty scary.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -935,6 +936,27 @@ class MinerImpl : public Mining
+        const MillisecondsDouble tick{1000};
+        auto now{std::chrono::steady_clock::now()};
+        auto deadline = now + timeout;
+        // std::chrono does not check against overflow
+        if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
```
It was in response to: https://github.com/bitcoin/bitcoin/pull/30409#discussion_r1713879417

ryanofsky replied:
Nice catch. I believe this can be simplified to:


```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -941,9 +941,7 @@ public:
         // Interrupt check interval
         const MillisecondsDouble tick{1000};
         auto now{std::chrono::steady_clock::now()};
-        auto deadline = now + timeout;
-        // std::chrono does not check against overflow
-        if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
+        const auto deadline{now + timeout};
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
             while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
```

This previous code written to handle the case where `now + timeout` calculation would overflow, basically assuming both values were unsigned integer values and overflow behavior would be well defined. But since `timeout` is a floating point number where "overflow" would just result in `inf` being set, trying to deal with it this way doesn't make sense.

theuni commented on src/node/kernel_notifications.h:
```diff
@@ -52,6 +56,12 @@ class KernelNotifications : public kernel::Notifications
+    Mutex m_tip_block_mutex;
+    std::condition_variable m_tip_block_cv;
+    //! The block for which the last blockTip notification was received for.
+    uint256 m_tip_block;
```
This needs a `GUARDED_BY(m_tip_block_mutex)`.

Adding it causes new warnings in `validation_chainstate_tests.cpp`. For ex:

> 
> src/test/validation_chainstate_tests.cpp:73:46: warning: reading variable 'm_tip_block' requires holding mutex 'm_node.notifications->m_tip_block_mutex' [-Wthread-safety-analysis]

I assume those particular accesses are harmless (maybe actually racy in the right conditions?), but future users may not be.

Related: It's a shame that the condvar/mutex leak out of here. Is there a reason not to add getter/waiter/notifier functions?

Sjors commented:
I'll work on a followup for the above things.

Sjors commented:
@maflcko #30966
